server/src/main/java/org/opensearch/extensions/rest/ExtensionRestResponse.java

```diff
@@ -0,0 +1,113 @@
+    private final boolean contentConsumed;
+
+    /**
+     * Creates a new response based on {@link XContentBuilder}.
+     *
+     * @param request  the REST request being responded to.
+     * @param status  The REST status.
+     * @param builder  The builder for the response.
+     */
+    public ExtensionRestResponse(ExtensionRestRequest request, RestStatus status, XContentBuilder builder) {
+        super(status, builder);
+        this.consumedParams = request.consumedParams();
+        this.contentConsumed = request.isContentConsumed();
+    }
+
+    /**
+     * Creates a new plain text response.
+     *
+     * @param request  the REST request being responded to.
+     * @param status  The REST status.
+     * @param content  A plain text response string.
+     */
+    public ExtensionRestResponse(ExtensionRestRequest request, RestStatus status, String content) {
+        super(status, content);
+        this.consumedParams = request.consumedParams();
+        this.contentConsumed = request.isContentConsumed();
+    }
+
+    /**
+     * Creates a new plain text response.
+     *
+     * @param request  the REST request being responded to.
+     * @param status  The REST status.
+     * @param contentType  The content type of the response string.
+     * @param content  A response string.
+     */
+    public ExtensionRestResponse(ExtensionRestRequest request, RestStatus status, String contentType, String content) {
+        super(status, contentType, content);
+        this.consumedParams = request.consumedParams();
+        this.contentConsumed = request.isContentConsumed();
+    }
+
+    /**
+     * Creates a binary response.
+     *
+     * @param request  the REST request being responded to.
+     * @param status  The REST status.
+     * @param contentType  The content type of the response bytes.
+     * @param content  Response bytes.
+     */
+    public ExtensionRestResponse(ExtensionRestRequest request, RestStatus status, String contentType, byte[] content) {
+        super(status, contentType, content);
+        this.consumedParams = request.consumedParams();
+        this.contentConsumed = request.isContentConsumed();
+    }
+
+    /**
+     * Creates a binary response.
+     *
+     * @param request  the REST request being responded to.
+     * @param status  The REST status.
+     * @param contentType  The content type of the response bytes.
+     * @param content  Response bytes.
+     */
+    public ExtensionRestResponse(ExtensionRestRequest request, RestStatus status, String contentType, BytesReference content) {
+        super(status, contentType, content);
+        this.consumedParams = request.consumedParams();
+        this.contentConsumed = request.isContentConsumed();
+    }
+
+    /**
+     * Gets the list of consumed parameters. These are needed to consume the parameters of the original request.
+     *
+     * @return the list of consumed params.
+     */
+    public List<String> getConsumedParams() {
```

Comment:
Trying to learn, how do regular HTTP servers respond? 
Is returning back consumed params a standard?

Reply:
I'm not sure how it is outside of OpenSearch.  All I know is that OpenSearch spits out an error if you include parameters which aren't included.

Reply:
Which is pretty interesting, Is there a code pointer in OpenSearch. I would love to see why we put it in place.

Reply:
Well, it's part of the [RestRequest](https://github.com/opensearch-project/OpenSearch/blob/3aef125d0dbf2f29d9bf00b84cb52892a27a5328/server/src/main/java/org/opensearch/rest/RestRequest.java#L86) class that we have to use to forward the relevant bits to the extension (and track the consuming).

Here's where it throws an exception on unconsumed params:
https://github.com/opensearch-project/OpenSearch/blob/60d7a098492c184e0b2a6fb0492d8c57b7626a54/server/src/main/java/org/opensearch/rest/BaseRestHandler.java#L105-L118

Reply:
Here's the PR from 6 years ago with more info:
https://github.com/opensearch-project/OpenSearch/commit/9a83ded55399c9e5af9de93f52125984e7e9c7cd